### PR TITLE
major refactoring

### LINKED ISSUE #14: incremental updates with exist triggers
Consider a collection managed by tuttle that has checked out commit A

- a later commit B creates new resources
- these resources are removed in commit C

When tuttle performs an incremental update from `A` to `C` it will first _create_ and then _delete_ those resources which might trigger events DOCUMENT_CREATE and DOCUMENT_REMOVE.

I would expect that these resources will be completely ignored / filtered in the aggregated action map.

### LINKED ISSUE #18: too many duplicate requests
At several points in the code the same request is performed to first check its return value and 
then perform the exact same request again to either get the error message or to parse the return value.

Example:

https://github.com/eeditiones/tuttle/blob/c58d66465eacb79d5f2147ce4ca34346e38a9d63/src/modules/github.xql#L22-L27

### PR DESCRIPTION
**Pull request by line-o**
fixes #14
fixes #18 

- Remove duplicate code
- make tuttle extensible to additional git services
- update node dependencies
- cleanup api code
- add unique operations (functions) for endpoints
- UI is now standards compliant and valid XHTML 
- test gitlab
- drastically reduce HTTP calls to remote services
- extend information that is exposed through /git/status
- refactor code for readability

### CODE REVIEW

File: src/index.html [hunk 1]
```diff
@@ -158,11 +159,11 @@ <h1><img class="tuttle" src="resources/images/HPTuttle-1866.png"></img>
                 <section class="login">
                     <fx-control ref="instance('auth')/user" update-event="enter">
                         <label>User</label>
-                        <input type="text" autofocus autocomplete="false">
+                        <input type="text" autofocus="" autocomplete="false" />
```
Comment (duncdrum) [inline src/index.html:162]: Shouldn t this be `autofocus="autofocus"` ?

File: src/modules/api.xql [hunk 2]
```diff
@@ -24,448 +23,433 @@ declare variable $api:definitions := ("api.json");
...
+            "message": $err:description,
+            "code": $err:code, "value": $err:value, 
+            "line": $err:line-number, "column": $err:column-number, "module": $err:module,
+            "request": map:remove($request, 'spec')
+         }
+    }
 };
 
 (:~ 
  : Trigger icremental update
```
Comment (duncdrum) [inline src/modules/api.xql:349]: Typo

File: src/modules/config.xql [hunk 3]
```diff
@@ -5,38 +5,50 @@ module namespace config="http://exist-db.org/apps/tuttle/config";
 (:~
  : Configurtion file
  :)
-declare variable $config:tuttle-config := doc("/db/apps/tuttle/data/tuttle.xml")/tuttle;
+declare variable $config:tuttle-config as element(tuttle) := doc("/db/apps/tuttle/data/tuttle.xml")/tuttle;
```
Comment (duncdrum) [inline src/modules/config.xql:8]: `//tuttle` ? Might be more robust with older versions of exist

File: src/modules/github.xql [hunk 4]
```diff
@@ -2,161 +2,135 @@ xquery version "3.1";
...
-            [app:shorten-sha(array:get($request, $size)?sha) , array:get($request, $size)?commit?message]
+declare function github:get-commits($config as map(*), $count as xs:integer?) as array(*)* {
+    let $url := github:commit-ref-url($config)
+    let $json := github:request-json($url, $config?token)
+    let $commits :=
+        if (empty($json))
+        then []
+        else if (empty($count))
+        then $json
+        else if ($count > array:size($json)) (: raise error here? returns everything :)
```
Comment (duncdrum) [inline src/modules/github.xql:80]: Looks like a good place for throwing an error

File: src/modules/gitlab.xql [hunk 5]
```diff
@@ -101,228 +65,218 @@ declare function gitlab:get-lastcommit-sha($config as map(*)) {
...
+declare function gitlab:get-commits($config as map(*), $count as xs:integer?) as array(*)* {
+    let $url := gitlab:commit-ref-url($config)
+    let $json := gitlab:request-json($url, $config?token)
+
+    let $commits :=
+        if (empty($json))                    (: raise error here? returns zero commits :)
+        then []
+        else if (empty($count))
+        then $json
+        else if ($count > array:size($json)) (: raise error here? returns everything :)
```
Comment (duncdrum) [inline src/modules/gitlab.xql:84]: S.a

**Review by duncdrum — CHANGES_REQUESTED** [review]
Looks very nice overall. Some small comments inline. 

Xar install needs fixing on ci

File: src/modules/api.xql (hunk 2 above)
Comment (line-o) [reply src/modules/api.xql:349]: Well yes apparently not mine

File: src/modules/config.xql (hunk 3 above)
Comment (line-o) [reply src/modules/config.xql:8]: I did not change the XPath so it will work as before

File: src/modules/github.xql (hunk 4 above)
Comment (line-o) [reply src/modules/github.xql:80]: Not necessarily. I just used 100 to hard-limit the size of returned elements. Think of a long commit history with thousands of commits.

**Comment by line-o** [comment]
Ah now we are hit with localhost DNS resolution on Linux as well. Fun.

File: src/index.html (hunk 1 above)
Comment (line-o) [reply src/index.html:162]: `autofocus="false"` works too :)

**Comment by duncdrum** [comment]
Nice. Should we adjust the sem-ver range in expath-pkg to excude exist v7, or is there something we can do about the failing tests and `:latest` ?

**Comment by line-o** [comment]
@duncdrum  I have found and fixed a couple of more issues in the meantime. Please wait with merging.

File: src/modules/github.xql (hunk 4 above)
Comment (line-o) [reply src/modules/github.xql]: The function will now throw only for numbers smaller than 1.

File: src/modules/gitlab.xql (hunk 5 above)
Comment (line-o) [reply src/modules/gitlab.xql]: This is intentionally not throwing.

File: src/modules/github.xql [hunk 6]
```diff
@@ -163,7 +179,7 @@ declare function github:aggregate-filechanges ($changes as map(*), $next as map(
 (:~ 
  : Run incremental update on collection in dry mode
  :) 
-declare function github:incremental-dry($config as map(*), $collection as xs:string) {
+declare function github:incremental-dry($config as map(*)) {
```
Comment (duncdrum) [inline src/modules/github.xql:182]: missing return type?

File: src/modules/github.xql [hunk 7]
```diff
@@ -174,11 +190,11 @@ declare function github:incremental-dry($config as map(*), $collection as xs:str
 (:~ 
  : Run incremental update on collection
  :) 
-declare function github:incremental($config as map(*), $collection as xs:string){
+declare function github:incremental($config as map(*)) {
```
Comment (duncdrum) [inline src/modules/github.xql:193]: here too?

File: src/modules/gitlab.xql [hunk 8]
```diff
@@ -214,61 +253,75 @@ declare %private function gitlab:incremental-delete($config as map(*), $collecti
...
+            }
         }
 };
 
 (:~
  : Incremental update fetch and add files from git
  :)
-declare %private function gitlab:incremental-add($config as map(*), $collection as xs:string, $sha as xs:string){
-    for $resource in gitlab:get-commit-files($config, $sha)?new
+declare %private function gitlab:incremental-add($config as map(*), $files as xs:string*, $sha as xs:string) {
```
Comment (duncdrum) [inline src/modules/gitlab.xql:267]: return type?

**Review by duncdrum — COMMENTED** [review]
looks great, and thanks for adding the gitlab tests, good thinking batman.

sadly, `:latest` is still failing

**Comment by line-o** [comment]
Regarding processor version: changing the dependency to
```
  <dependency processor="http://exist-db.org" semver-min="5.5.1" semver-max="6" />
```
should do, but I think we should postpone this until we are about to release 2.0.0

**Comment by duncdrum** [comment]
@line-o i m ok with that, but would prefer to at least mention the known incompatibility on the readme, in that case.

**Comment by line-o** [comment]
It is worth to put that info somewhere. Since this is the case for the current release and main this can be done in a separate PR.

**Review by JoernT — APPROVED** [review]

**Pull request merged**